src/types/roots.rs

```diff
@@ -263,76 +247,54 @@ impl RootFdfSolver {
     /// This function initializes, or reinitializes, an existing solver s to use the function and
     /// derivative fdf and the initial guess root.
     #[doc(alias = "gsl_root_fdfsolver_set")]
-    pub fn set<F: Fn(f64) -> f64, DF: Fn(f64) -> f64, FDF: Fn(f64, &mut f64, &mut f64)>(
+    pub fn set<
+        'a,
+        F: Fn(f64) -> f64 + 'a,
+        DF: Fn(f64) -> f64 + 'a,
+        FDF: Fn(f64, &mut f64, &mut f64),
+    >(
         &mut self,
         f: F,
         df: DF,
         fdf: FDF,
         root: f64,
     ) -> ::Value {
-        unsafe extern "C" fn inner_f<F: Fn(f64) -> f64>(
-            x: c_double,
-            params: *mut c_void,
-        ) -> c_double {
-            let params: &(*const F, *const (), *const ()) =
-                &*(params as *const (*const F, *const (), *const ()));
-            let f = &*params.0;
+        // convert rust functions to C
+        unsafe extern "C" fn inner_f<'a, F: Fn(f64) -> f64 + 'a>(
```

Comment:
You need to store `f`, `df` and `fdf` here as well, otherwise they will be called outside of this scope.